bgpd/bgp_packet.c

```diff
@@ -1005,10 +1005,35 @@ static int bgp_collision_detect(struct peer *new, struct in_addr remote_id)
 					return -1;
 				}
 			else {
-				if (ntohl(peer->local_id.s_addr) ==
-				    ntohl(remote_id.s_addr))
-					flog_err(EC_BGP_ROUTER_ID_SAME, "Peer's router-id %s is the same as ours",
-						 inet_ntoa(remote_id));
+				/* rfc6286:
+				 * If the BGP Identifiers of the peers
+				 * involved in the connection collision
+				 * are identical, then the connection
+				 * initiated by the BGP speaker with the
+				 * larger AS number is preserved.
+				 */
+				if (ntohl(peer->local_id.s_addr)
+					    == ntohl(remote_id.s_addr)
+				    && peer->local_as < peer->as) {
+					flog_err(
+						EC_BGP_ROUTER_ID_SAME,
+						"Peer's router-id %s is the same as ours",
+						inet_ntoa(remote_id));
+
+					if (!CHECK_FLAG(
+						    peer->sflags,
+						    PEER_STATUS_ACCEPT_PEER)) {
+						bgp_notify_send(
+							peer, BGP_NOTIFY_CEASE,
+							BGP_NOTIFY_CEASE_COLLISION_RESOLUTION);
+						return 1;
+					} else {
+						bgp_notify_send(
+							new, BGP_NOTIFY_CEASE,
+							BGP_NOTIFY_CEASE_COLLISION_RESOLUTION);
+						return -1;
+					}
+				}
 
 				/* 3. Otherwise, the local system closes newly
```

Comment:
The next 2 checks need to be not executed if your newly added code is executed, otherwise it's possible both ends of the connection will be closed.  (Or neither.)

I would recommend changing the "1." if condition from
```
                        if (ntohl(peer->local_id.s_addr)
                            < ntohl(remote_id.s_addr))
```
to
```
                        if (ntohl(peer->local_id.s_addr)
                            < ntohl(remote_id.s_addr)
                    || (peer->local_id.s_addr == remote_id.s_addr && peer->local_as < peer->as))
```
and move the log message out of the condition under a new `else if (local == remote && peer_as == local_as)` block.

Reply:
My brain failed to process the `return` statements there.  I would still change the code structure from

```
/* 1. */
if (ip < ip) {
  /* 2. */
  A
} else {
  if (ip == ip && as < as) {
    /* new */
    log()
    A
  }
  /* 3. */
  B
}
```
to
```
/* 1. */
if (ip < ip || (ip == ip && as < as)) {
  /* 2. */
  A
} else {
  if (ip == ip && as == as)
    log();
  /* 3. */
  B
}
```

Reply:
Thanks, I'll fix ;-)

Reply:
@eqvinox updated :1st_place_medal: